### DISCUSSION TITLE
generating complementary PWM instead of DAC using the audio data

### DISCUSSION THREAD

**Discussion by tesla-cat** [pschatzmann/ESP32-A2DP#452] [Q&A] [answered]
**after getting the data, how can I generate PWM using it instead of DAC?**
**I want to make a power supply whose output is modulated by the audio signal, what i am thinking is using a half bridge inverter topology, therefore I need two complementary PWM signals (PWM signals where one has 180 degree phase lag from another), and the duty cycle of the PWM signal should be proportional to the audio signal**

> Or you can access the packet:

```c++
// In the setup function:
a2dp_sink.set_stream_reader(read_data_stream);

// Then somewhere in your sketch:
void read_data_stream(const uint8_t *data, uint32_t length)
{
  int16_t *samples = (int16_t*) data;
  uint32_t sample_count = length/2;
  // Do something with the data packet
}
```

**Comment by pschatzmann** [comment, marked as answer]
Just use the AudioTools project (as suggested in the documentation). 
In the basic-api-exmples you can find an [a2dp example which outputs to i2s](https://github.com/pschatzmann/arduino-audio-tools/blob/main/examples/examples-basic-api/basic-a2dp-i2s/basic-a2dp-i2s.ino). Just use the https://pschatzmann.github.io/arduino-audio-tools/classaudio__tools_1_1_p_w_m_audio_output.html as output class instead.

**Reply by tesla-cat** [reply]
is this able to generate complementary PWM i was talking about?

**Comment by pschatzmann** [comment]
That would be your starting point: Not sure what the best approach for the phase shift is: you might want to implement this using some digital filters.  There is a chapter in the Wiki.

**Reply by tesla-cat** [reply]
i tried the following code, and expect to see the led lighting up on the ESP32 board (pin 2), but i dont see anything

```c++
#include "AudioTools.h"
#include "BluetoothA2DPSink.h"

BluetoothA2DPSink a2dp_sink;
PWMAudioStream pwm;

void read_data_stream(const uint8_t *data, uint32_t length) {
    pwm.write(data, length);
}

void setup() {
    Serial.begin(115200);
    AudioLogger::instance().begin(Serial, AudioLogger::Warning);

    a2dp_sink.set_stream_reader(read_data_stream, false);

    a2dp_sink.set_auto_reconnect(false);
    a2dp_sink.start("a2dp-i2s");

    auto cfg = pwm.defaultConfig();
    // cfg.pin_data = 23;
    cfg.sample_rate = a2dp_sink.sample_rate();
    cfg.channels = 2;
    cfg.bits_per_sample = 16;
    // cfg.buffer_count = 8;
    cfg.buffer_size = 256;
    int pins[] = {2, 23};
    cfg.setPins(pins);
    pwm.begin(cfg);
}

void loop() { delay(100); }
```

**Reply by pschatzmann** [reply]
I can't reproduce your issue: After connecting and starting to stream some audio the LED lights up as expected.

**Comment by pschatzmann** [comment]
First thought: did you adjust the volume on your phone.
Build your sketch step by step and test each step. Thats the only way to identify where your issue is.
Here are some hints: https://github.com/pschatzmann/arduino-audio-tools/wiki/It's-not-working
To analyse A2DP you can activate the logging.
There are some sketches using PWM output: i have tested them recently with some earphones.

**Comment by pschatzmann** [comment]
Just one more question: I don't quite understand your statement 'power supply whose output is modulated by the audio signal'. Do you really need to have a resolution of 44100 updates per second to adjust your power supply ?

**Reply by tesla-cat** [reply]
i need a power supply that generates a voltage between 0 to 300 volts, with the same shape as my audio signal. 
when there is no audio, the power supply outputs 0 volts, when the audio reaches its peak amplitude, the power supply outputs 300 volts.

**Reply by pschatzmann** [reply]
I have the feeling that in this case it might be just better to measure the volume and update your power supply at a much lower rate. E.g after each 1024 samples. I suggest that you also have a look the VolumeOutput examples. This also gives you a much higher resolution.

**Reply by tesla-cat** [reply]
if you are interested, what i am trying to build is something called musical QCW Tesla Coil

https://www.youtube.com/watch?v=R18nwlPJxzc

https://www.youtube.com/watch?v=pRddSByLmKM

https://www.youtube.com/watch?v=ovFdAaDbyc4

**Comment by pschatzmann** [comment]
This confirms my suspicion: I think the low resolution, high sampling rate approach of the PWMStream does not help you at all.
I would go for the volume approach. 

You want to update your voltage based on the actual volume and not based on the individual audio waves!

**Reply by tesla-cat** [reply]
okay, thanks.

FYI, this is the best video explaining what i am trying to do:  

https://www.youtube.com/watch?v=QpKNbvonSic&t=1s

**Reply by pschatzmann** [reply]
Hmm, this leaves quite some questions open: I would suggest to follow both of my suggested tracks. The video suggests that with higher sampling rates you get better shape control of the sparks. 

I still think that 44100 might be a bit high for your purpose and I suggest that you look into the topic of downsampling.
I am not sure about the impact of noise. The best would be to use something with a proper Audio DAC like the AudioKit board or any external DAC as suggested in the WIKI
